addons/product/models/product.py

```diff
@@ -223,9 +223,22 @@ def init(self):
         self.env.cr.execute("CREATE UNIQUE INDEX IF NOT EXISTS product_product_combination_unique ON %s (product_tmpl_id, combination_indices) WHERE active is true"
             % self._table)
 
-    _sql_constraints = [
-        ('barcode_uniq', 'unique(barcode)', "A barcode can only be assigned to one product !"),
-    ]
+    # TODO: remove before merge
+    # _sql_constraints = [
+    #     ('barcode_uniq', 'unique(barcode)', "A barcode can only be assigned to one product !"),
+    # ]
+
+    @api.constrains("barcode")
+    def _check_barcode_unique(self):
+        products_with_same_barcode = self.env['product.product'].sudo().search([('barcode', '=', self.barcode)]) - self
+        if products_with_same_barcode:
+            raise ValidationError(
+                _("This barcode is already in use by:\n\n{products_with_same_barcode}").format(
+                    products_with_same_barcode="\n".join(
+                        [product.name for product in products_with_same_barcode]
```

Comment:
`product.name` --> `product.display_name`
Use display_name in case of variante.

Reply:
Thank you, I wasn't aware of that one!